Makefile

```diff
@@ -89,47 +87,77 @@ ifneq ($(CHECKSTYLE),0)
 	PERL5LIB=lib/perlcritic:$$PERL5LIB perlcritic lib
 endif
 
-.PHONY: test
-ifeq ($(TRAVIS),true)
-test: run-tests-within-container
-else
-test: checkstyle test-with-database
-endif
+.PHONY: phony_explicit
+phony_explicit: 
+
+.PHONY: test-with-db
+test-with-db: test-with-db-default
 
-.PHONY: test-unit-and-integration
-test-unit-and-integration:
-	OPENQA_CONFIG= prove ${PROVE_LIB_ARGS} ${PROVE_ARGS}
+test-with-db%: phony_explicit
+	@$(MAKE) $(subst test-with-db,test-with-database,$@)
 
 .PHONY: test-with-database
-test-with-database:
+test-with-database: test-with-database-default
+
+test-with-database%: phony_explicit
 	test -d $(TEST_PG_PATH) && (pg_ctl -D $(TEST_PG_PATH) -s status >&/dev/null || pg_ctl -D $(TEST_PG_PATH) -s start) || ./t/test_postgresql $(TEST_PG_PATH)
-	$(MAKE) test-unit-and-integration TEST_PG="DBI:Pg:dbname=openqa_test;host=$(TEST_PG_PATH)"
+	$(MAKE) $(subst -with-database,,$@) TEST_PG="DBI:Pg:dbname=openqa_test;host=$(TEST_PG_PATH)"
 	-pg_ctl -D $(TEST_PG_PATH) stop
 
-# prepares running the tests within Docker (eg. pulls os-autoinst) and then runs the tests considering
-# the test matrix environment variables
-# note: This is supposed to run within the Docker container unlike `launch-docker-to-run-tests-within`
-#       which launches the container.
-.PHONY: run-tests-within-container
-run-tests-within-container:
-	script/run-tests-within-container
+.PHONY: test
+test:
+	@[ -z ${PERL5OPT} ] || echo PERL5OPT=${PERL5OPT}
+	prove ${PROVE_LIB_ARGS} ${PROVE_ARGS}
```

Comment:
The problem I see here is that previously "make test" would start "checkstyle" and "test-with-database", now you are doing something different. Making this a new meaning is at least misleading and I do not see the motivation for making it harder again for users to not be able to directly use `make test` without the database setup. Also "checkstyle" would not be called anymore. I would prefer if you can change this back.

Reply:
That is not true for CI, where `make test` would not start DB and would run `checkstyle` only if UITESTS is not set to 0 - do you want to stick to this logic? https://github.com/os-autoinst/openQA/blob/bb62d7666521acd8a0ff57e1b7c65d3f49ae9a55/script/run-tests-within-container#L54

Reply:
well, https://github.com/os-autoinst/openQA/blob/bb62d7666521acd8a0ff57e1b7c65d3f49ae9a55/script/run-tests-within-container#L54 calls an explicit low-level make target which is not "make test". I don't see the direct relation. I just meant to keep the meaning of "make test" as the generic, common entry path for testing which is valid for users as in before, i.e. call it and it does something useful without needing further setup.

Reply:
But that whole script is called from inside `make test` https://github.com/os-autoinst/openQA/blob/master/Makefile#L94. Thus `checkstyle` is run from `make test` only if UITESTS=0.
Overall I don't see any advantage of starting DB and performing `checkstyle` by default for those who doesn't care about what exactly they want to test at cost of requiring to carry over CHECKSTYLE=0 and potential implications of always starting DB for those who cares. E.g. what is purpose of starting DB and running `checkstyle` as part of command `PROVE_ARGS=t/00* make test` or any other test which doesn't use DB?

Reply:
Obviously there is no use in starting a database to just check the style.

Maybe I still made myself not clear (or you have understood me already and want to argue about something different) so let me give you a very simple example. If anyone asks: "What do I need to call to run all openQA tests locally?", the answer can currently be "Call 'make test'". This should stay. This is much more about what https://github.com/os-autoinst/openQA/blob/master/Makefile#L96 did, not https://github.com/os-autoinst/openQA/blob/master/Makefile#L94

Reply:
Obviously the answer to that question will be "it dependes", because if you currently do "make test" inside Travis it will e.g. not run t/ui/*.t tests or will not run `scheckstyle` if `UITESTS` is set to 1. I am not sure why you ignore my points of inconsistency of current behavior and of useless DB start when `make test` is called with predefined PROVE_ARGS (which don't require database) and just repeat the same over and over. And instead of making the command consistent across all cases you concentrate on single use case, which is not even official Makefile standard https://www.gnu.org/prep/standards/html_node/Standard-Targets.html#Standard-Targets and e.g. not used in spec file (which uses `make test-with-database`)
Even our documentation is inconsistent, suggesting to use `make test` for all tests and `prove` for individual test cases, which is conceptually weak. Let's concentrate on improving  current behavior and make it consistent rather than random points which you don't like and ignore everything else.
If you insist on following official guidelines - let's add `make check` target which will just call `make test-with-database` - will it work for you?

Reply:
I wanted to "ignore your points" regarding CI use because it feels like we are getting nowhere so I wanted to focus on the "entry point for humans" only.

> let's add make check target which will just call make test-with-database - will it work for you?

If you must insist that you need to change this much of independant things at the same time – which I still do not understand why, I would prefer to keep the meaning of "make test" for humans the same as in before – then your `make check` target is the path of least resistance :)

Reply:
> I wanted to focus on the "entry point for humans" only

Don't humans write .travis.yml? All these things are very dependent - I couldn't reuse code in config.yml without adding proper make targets and couldn't test without excluding out unstable tests and it is inconsistent to use `CHECKSTYLE=0` in every circleci job except one.
I don't see much benefit of testing Makefile changes in Travis first, but it for sure will be much more time consuming
But I agree that `make check` is needed, so I've added it.

Reply:
> but it for sure will be much more time consuming

I agree and it is likely not worth it.

Reply:
I give up. I hope you will at least adjust the documentation.